### PR TITLE
Work on a hero image for the website, including a typing animation.

### PR DESCRIPTION
**Pull request by k-sharpe**

### CODE REVIEW

File: portfolio/src/main/webapp/index.html [hunk 1]
```diff
@@ -6,7 +6,11 @@
     <link rel="stylesheet" href="style.css">
     <script src="script.js"></script>
   </head>
-  <body>
+  <body onload="typing()">
+    <div class="Landing">
```
Comment (andreahu12) [inline portfolio/src/main/webapp/index.html:10]: nit: I think the style guide says to use lowercase in HTML and CSS. go/htmlcssstyle#Capitalization

File: portfolio/src/main/webapp/style.css [hunk 2]
```diff
@@ -3,12 +3,26 @@
...
-    width: 0px;
+
+.Landing {
+    background-image: url("images/dog.png");
+    height: 400px;
+    width: 100%;
+    background-position: center;
+    background-repeat: no-repeat;
+    background-size: cover;
+    position: relative;
```
Comment (andreahu12) [inline portfolio/src/main/webapp/style.css:14]: ditto nit: I think the style guide says to use lowercase in HTML and CSS. go/htmlcssstyle#Capitalization

File: portfolio/src/main/webapp/script.js [hunk 3]
```diff
@@ -26,3 +26,25 @@ function addRandomGreeting() {
...
+
+// Delay in milliseconds.
+var speed = 150;
+
+/**
+ * Function that handles the actual typewriter effect.
+ */
+function typing() {
+  if (typewriterStringIndex < txt.length) {
+      typingDestination = document.getElementById('TypewriterText')
```
Comment (andreahu12) [inline portfolio/src/main/webapp/script.js:44]: Make sure to end lines with semicolon ;
go/htmlcssstyle#Declaration_Stops

File: portfolio/src/main/webapp/script.js [hunk 4]
```diff
@@ -26,3 +26,25 @@ function addRandomGreeting() {
...
+      charToAdd = txt.charAt(typewriterStringIndex)
```
Comment (andreahu12) [inline portfolio/src/main/webapp/script.js:45]: ditto: go/htmlcssstyle#Declaration_Stops

File: portfolio/src/main/webapp/script.js [hunk 5]
```diff
@@ -26,3 +26,25 @@ function addRandomGreeting() {
   const greetingContainer = document.getElementById('greeting-container');
   greetingContainer.innerText = greeting;
 }
+
+/**
+ * Effect of typing on the hero.
+ */
+var typewriterStringIndex = 0;
```
Comment (andreahu12) [inline portfolio/src/main/webapp/script.js:33]: I think people generally use let instead of var because the scope is more expected?
https://stackoverflow.com/questions/762011/whats-the-difference-between-using-let-and-var/11444416#11444416

File: portfolio/src/main/webapp/script.js [hunk 6]
```diff
@@ -26,3 +26,25 @@ function addRandomGreeting() {
...
+var txt = 'My Portfolio';
+
+// Delay in milliseconds.
+var speed = 150;
+
+/**
+ * Function that handles the actual typewriter effect.
+ */
+function typing() {
+  if (typewriterStringIndex < txt.length) {
```
Comment (andreahu12) [inline portfolio/src/main/webapp/script.js:43]: Could call it typingAnimation() for clarity?

Comment (k-sharpe) [reply portfolio/src/main/webapp/script.js]: Sure, made all of my variables named better

File: portfolio/src/main/webapp/script.js (hunk 5 above)
Comment (k-sharpe) [reply portfolio/src/main/webapp/script.js]: Sgtm

File: portfolio/src/main/webapp/index.html [hunk 7]
```diff
@@ -6,7 +6,11 @@
...
+        <div id="TypewriterText" class="TypewriterText">
+        </div>
```
[the remark this answers is not in the document]
Comment (k-sharpe) [reply portfolio/src/main/webapp/index.html]: Should I be using double spacing on my .js files too? I know two space applies for html and css but am unsure if it's the case for java script as well. Thank you

File: portfolio/src/main/webapp/script.js [hunk 8]
```diff
@@ -26,3 +26,63 @@ function addRandomGreeting() {
...
+ */
+
+// Index on a given string to type next.
+let typewriterStringIndex = 0;
+
+// Index of the current word
+let typePhrase = 0;
+
+// All texts to type
+let typewriterText = ['My Passions', 'My Projects', 'My Portfolio'];
```
Comment (jgao23) [inline portfolio/src/main/webapp/script.js:41]: i would do something like the following:


> const typewriterText = ['My Passions', 'My Projects', 'My Portfolio'].map(text => text.replace(' ', '\xa0'));
and move the comment in cleanTypewriterText() to here. That way the text is still readable but you don't need to clean the text every time you call typing().

File: portfolio/src/main/webapp/script.js [hunk 9]
```diff
@@ -26,3 +26,63 @@ function addRandomGreeting() {
...
+}
+
+/**
+ * Cleans the typewriter text to make it display properly.
+ * Why? a normal space ' ' won't be displayed, must use
+ * '\xa0'.
+ */
+ function cleanTypewriterText() {
+     for (let i = 0; i < typewriterText.length; i++) {
+        typewriterText[i] = typewriterText[i].replace(' ', '\xa0');
```
Comment (jgao23) [inline portfolio/src/main/webapp/script.js:86]: why doesn't just the ' ' work?

File: portfolio/src/main/webapp/script.js [hunk 10]
```diff
@@ -26,3 +26,63 @@ function addRandomGreeting() {
...
+
+// Delay in milliseconds.
+let speed = 150;
+
+/**
+ * Function that handles the actual typewriter effect.
+ */
+function typing() {
+    cleanTypewriterText();
+    var typingDestination = document.getElementById('typewriter-text');
```
Comment (jgao23) [inline portfolio/src/main/webapp/script.js:51]: i would make this a global const like so:

> const typingElement = document.getElementById('typewriter-text');

So you don't need to redefine in reverseTyping and typing()

File: portfolio/src/main/webapp/index.html [hunk 11]
```diff
@@ -6,7 +6,11 @@
     <link rel="stylesheet" href="style.css">
     <script src="script.js"></script>
   </head>
-  <body>
+  <body onload="typing()">
+    <div class="landing">
+      <div id="typewriter-text" class="typewriter-text">
```
Comment (jgao23) [inline portfolio/src/main/webapp/index.html:11]: don't need to set both id and class here. looks like the starter code just sets the id, so i would follow that pattern.

File: portfolio/src/main/webapp/script.js [hunk 12]
```diff
@@ -26,3 +26,63 @@ function addRandomGreeting() {
...
+    if (typewriterStringIndex < typewriterText[typePhrase].length) {
+      charToAdd = typewriterText[typePhrase].charAt(typewriterStringIndex);
+      typingDestination.innerText += charToAdd;
```
Comment (jgao23) [inline portfolio/src/main/webapp/script.js:54]: you used substr in reverseType(), why not do it here as well?

File: portfolio/src/main/webapp/script.js [hunk 13]
```diff
@@ -26,3 +26,63 @@ function addRandomGreeting() {
...
+
+/**
+ * Effect of typing on the hero.
+ */
+
+// Index on a given string to type next.
+let typewriterStringIndex = 0;
+
+// Index of the current word
+let typePhrase = 0;
```
Comment (jgao23) [inline portfolio/src/main/webapp/script.js:38]: since you modify these values, i would just define them in typing()

File: portfolio/src/main/webapp/script.js [hunk 14]
```diff
@@ -26,3 +26,63 @@ function addRandomGreeting() {
...
+
+// All texts to type
+let typewriterText = ['My Passions', 'My Projects', 'My Portfolio'];
+
+// Delay in milliseconds.
+let speed = 150;
```
Comment (jgao23) [inline portfolio/src/main/webapp/script.js:44]: it looks like you adjust speed in different areas. i would either just define separate const for the different speeds something like: 

> const DELETE_DELAY = 50;
> const ADD_DELAY = 150;
> const WORD_DELAY = 75;

or i would just hard code the values in.

File: portfolio/src/main/webapp/script.js [hunk 15]
```diff
@@ -26,3 +26,63 @@ function addRandomGreeting() {
...
+ * Adjustments to speed are purely stylistic and will not break anything.
+ */
+function reverseTyping() {
+    var typingDestination = document.getElementById('typewriter-text');
+    if (typingDestination.innerText.length === 0) {
+        setTimeout(typing, speed * 2);
+    } else {
+        let lenOfDestText = typingDestination.innerText.length;
+        typingDestination.innerText = typingDestination.innerText.substr(0, lenOfDestText - 1);
+        setTimeout(reverseTyping, speed / 3);
```
Comment (jgao23) [inline portfolio/src/main/webapp/script.js:75]: noice recursion, can you do that for typing forward too so you don't need to keep state? you can also try doing both iteratively (2 for loops), but i know it might get complex because of the weird scoping with setTimeout

File: portfolio/src/main/webapp/script.js [hunk 16]
```diff
@@ -26,3 +26,63 @@ function addRandomGreeting() {
...
+// All texts to type
+let typewriterText = ['My Passions', 'My Projects', 'My Portfolio'];
+
+// Delay in milliseconds.
+let speed = 150;
+
+/**
+ * Function that handles the actual typewriter effect.
+ */
+function typing() {
```
Comment (jgao23) [inline portfolio/src/main/webapp/script.js:49]: i would maybe rename this to renderTitle(). also, i think it would be cleaner to split out the type forward part into a separate helper function since it would be easier to read.

File: portfolio/src/main/webapp/style.css [hunk 17]
```diff
@@ -3,12 +3,27 @@
...
+.typewriter-text {
+    top: 50%;
+    left: 50%;
+    font-family: verdana, helvetica, sans-serif;
+    color: coral;
+    text-align: center;
+    position: absolute;
+    font-size: 60px;
+    transform: translate(-50%, -50%);
+}
```
Comment (jgao23) [inline portfolio/src/main/webapp/style.css:26]: indentation is incorrect

File: portfolio/src/main/webapp/index.html (hunk 7 above)
[the remark this answers is not in the document]
Comment (jgao23) [reply portfolio/src/main/webapp/index.html]: Follow the starter code pattern. If the starter code is 2 spaces in the JS file, do that too.

File: portfolio/src/main/webapp/script.js (hunk 9 above)
Comment (k-sharpe) [reply portfolio/src/main/webapp/script.js:86]: The thing i added was a non-breaking whitespace. On my local dev server without it the typing function would output "MyPortfolio" instead of "My Portfolio" which was fixed by the \xa0

Comment (jgao23) [reply portfolio/src/main/webapp/script.js:86]: right but why do we need to use \xa0 when in other places the space just works as is?

File: portfolio/src/main/webapp/script.js (hunk 10 above)
Comment (k-sharpe) [reply portfolio/src/main/webapp/script.js:51]: When I do that, typingElement becomes null, I think because the page hasn't loaded it yet.

File: portfolio/src/main/webapp/script.js (hunk 13 above)
Comment (andreahu12) [reply portfolio/src/main/webapp/script.js:38]: also could be called typewriterTextIndex for clarity since it's the index for typewriterText

File: portfolio/src/main/webapp/script.js (hunk 9 above)
Comment (k-sharpe) [reply portfolio/src/main/webapp/script.js:86]: This is what I found: https://stackoverflow.com/questions/47768523/empty-spaces-are-ignored-by-the-innertext-property
TL;DR: I read from inner text which erases white space, and then add a single ' ', infinitely looping.

File: portfolio/src/main/webapp/style.css [hunk 18]
```diff
@@ -3,12 +3,27 @@
...
+#typewriter-text {
+    top: 50%;
+    left: 50%;
+    font-family: verdana, helvetica, sans-serif;
+    color: coral;
+    text-align: center;
+    position: absolute;
+    font-size: 60px;
+    transform: translate(-50%, -50%);
+}
```
Comment (jgao23) [inline portfolio/src/main/webapp/style.css:26]: fix indentation

File: portfolio/src/main/webapp/script.js [hunk 19]
```diff
@@ -26,3 +26,39 @@ function addRandomGreeting() {
   const greetingContainer = document.getElementById('greeting-container');
   greetingContainer.innerText = greeting;
 }
+
+// Easy way of keeping track of what word to print.
+let typewriterTextIndex = 0;
```
Comment (jgao23) [inline portfolio/src/main/webapp/script.js:31]: a better name might be currentTypewriterTextIndex. And also can be defined in the renderHeroText() func itself. global variables are usually immutable

File: portfolio/src/main/webapp/script.js [hunk 20]
```diff
@@ -26,3 +26,39 @@ function addRandomGreeting() {
...
+
+// Easy-access list to make editing the hero text easy. 
+let typewriterText = ['My Passions', 'My Projects', 'My Portfolio'];
```
Comment (jgao23) [inline portfolio/src/main/webapp/script.js:34]: this should be a const now since you should never be reassigning typewriterText

File: portfolio/src/main/webapp/script.js [hunk 21]
```diff
@@ -26,3 +26,39 @@ function addRandomGreeting() {
...
+/**
+ * A helper function complimentary of the type-forward function to animate the home page.
+ * Adjustments to speed are purely stylistic and will not break anything.
+ */
+function reverseHeroText() {
+  if (typingDestination.innerText.length === 0) {
+      setTimeout(renderHeroText, 300);
+  } else {
+      let lenOfDestText = typingDestination.innerText.length;
+      typingDestination.innerText = typingDestination.innerText.substr(0, lenOfDestText - 1);
```
Comment (jgao23) [inline portfolio/src/main/webapp/script.js:61]: supernit: keep it consistent; you did currentCharIndex++ above, so maybe you should rename lenOfDestText to currentCharIndex and do currentCharIndex--

File: portfolio/src/main/webapp/script.js (hunk 19 above)
Comment (k-sharpe) [reply portfolio/src/main/webapp/script.js]: Fixed, also got it out of global

File: portfolio/src/main/webapp/script.js (hunk 20 above)
Comment (k-sharpe) [reply portfolio/src/main/webapp/script.js]: Fixed this as well, I think I just now finally understand how variables work in JS

File: portfolio/src/main/webapp/script.js [hunk 22]
```diff
@@ -17,7 +17,7 @@
  */
 function addRandomGreeting() {
   const greetings =
-      ["I have a California Real Estate Salesperson Licence!", "My dog is forever a  puppy", "I work at Google!", "I played water polo in high school"];
+    ["I have a California Real Estate Salesperson Licence!", "My dog is forever a  puppy", "I work at Google!", "I played water polo in high school"];
```
Comment (andreahu12) [inline portfolio/src/main/webapp/script.js:20]: nit: I think it's spelled `License`

Comment (andreahu12) [inline portfolio/src/main/webapp/script.js:20]: nit: `"My dog is forever a  puppy"` has an extra space between `a` and `puppy`

File: portfolio/src/main/webapp/script.js [hunk 23]
```diff
@@ -26,3 +26,36 @@ function addRandomGreeting() {
...
+ * Animates the front page.
+ * @param {number} currentCharIndex Keeps track of the next letter to add to the element. 0 by default.
+ * @param {number} curTypewriterTextIndex Keeps track of the current word in typewriterText.
+ */
+function renderHeroText(currentCharIndex = 0, curTypewriterTextIndex = 0) {
+  typingDestination = document.getElementById("typewriter-text");
+  if (currentCharIndex <= typewriterText[curTypewriterTextIndex].length) {
+    typingDestination.innerText = typewriterText[curTypewriterTextIndex].substr(0, currentCharIndex);
+    setTimeout(renderHeroText, 150, ++currentCharIndex, curTypewriterTextIndex);
+  } else if (curTypewriterTextIndex < typewriterText.length - 1) {
```
Comment (andreahu12) [inline portfolio/src/main/webapp/script.js:43]: Let's add parameter name comments to make it clearer what raw values represent
https://google.github.io/styleguide/jsguide.html#formatting-param-name-comments

`setTimeout(renderHeroText, /* milliseconds= */ 150, ++currentCharIndex, curTypewriterTextIndex);`

File: portfolio/src/main/webapp/script.js [hunk 24]
```diff
@@ -26,3 +26,36 @@ function addRandomGreeting() {
...
+
+/**
+ * A helper function complimentary of the type-forward function to animate the home page.
+ * Adjustments to speed are purely stylistic and will not break anything.
+ * @param {number} currentCharIndex Keeps track of the next letter to remove from the element.
+ * @param {number} curTypewriterTextIndex Keeps track of current word in typewriterText
+ */ 
+function reverseHeroText(currentCharIndex, curTypewriterTextIndex) { 
+  if (typingDestination.innerText.length === 0) {
+    setTimeout(renderHeroText, 300, 0, curTypewriterTextIndex);
```
Comment (andreahu12) [inline portfolio/src/main/webapp/script.js:56]: Let's add parameter name comments to make it clearer what raw values represent
https://google.github.io/styleguide/jsguide.html#formatting-param-name-comments

`setTimeout(renderHeroText, /* milliseconds= */ 300, /* currentCharIndex= */ 0, curTypewriterTextIndex);`

File: portfolio/src/main/webapp/script.js [hunk 25]
```diff
@@ -26,3 +26,36 @@ function addRandomGreeting() {
...
+  } else {
+    typingDestination.innerText = typingDestination.innerText.substr(0, currentCharIndex);
+    setTimeout(reverseHeroText, 50, --currentCharIndex, curTypewriterTextIndex);
```
Comment (andreahu12) [inline portfolio/src/main/webapp/script.js:59]: Let's add parameter name comments to make it clearer what raw values represent
https://google.github.io/styleguide/jsguide.html#formatting-param-name-comments

`setTimeout(renderHeroText, /* milliseconds= */ 50, --currentCharIndex, curTypewriterTextIndex);`

**Comment by k-sharpe** [comment]
Woah I did not intend for all that merging to happen, but the only new changes are the typos and the parameter names getting modified.

**Comment by jgao23** [comment]
Good stuff!
